<a href="https://colab.research.google.com/github/HimanshuSahoo/Graph-Based-Recommendation-System-in-Python/blob/main/2_Data_Exploration_and_Data_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip3 install duckdb==0.3.1

### About DuckDB [Official Webpage](https://duckdb.org/) | [Query Parquet with DuckDB](https://duckdb.org/2021/06/25/querying-parquet.html)

In [ ]:
import pandas as pd
import numpy as np
import duckdb
import time
con = duckdb.connect(database=':memory:', read_only=False)

## Exploration Query

In [ ]:
%%time
con.query('''
SELECT event_type,
       count(*) as event_count
FROM '../Data/optimised_raw_data.parquet'
GROUP BY 1''').df()

CPU times: user 1.8 s, sys: 68.9 ms, total: 1.87 s
Wall time: 2.09 s


event_type  event_count
0       view     52490785
1       cart      2641249
2   purchase       835007

In [ ]:
con.query('''
SELECT *
FROM '../Data/optimised_raw_data.parquet'
WHERE event_type = 'view'
LIMIT 5''').df()

event_time event_type  product_id          category_id  \
0  2020-01-01 00:00:00 UTC       view     1005073  2232732093077520756   
1  2020-01-01 00:00:01 UTC       view     1005192  2232732093077520756   
2  2020-01-01 00:00:01 UTC       view   100063693  2053013552427434207   
3  2020-01-01 00:00:01 UTC       view     5100816  2232732103831716449   
4  2020-01-01 00:00:02 UTC       view   100014325  2232732103294845523   

              category_code    brand        price    user_id  \
0  construction.tools.light  samsung  1130.020020  519698804   
1  construction.tools.light    meizu   205.669998  527767423   
2             apparel.shirt   turtle   136.429993  519046195   
3             apparel.shoes   xiaomi    29.950001  518269232   
4    apparel.shoes.step_ins    intel   167.199997  587748686   

                           user_session  
0  69b5d72f-fd6e-4fed-aa23-1286b2ca89a0  
1  7f596032-ccbf-4643-9bad-e36a209512b4  
2  d1e2f343-84bb-49bd-b13d-ca0f1ed9910e  
3  0444841c-38ef-410c-b11f-7b35ea4e5991  
4  31b7d4cf-dfac-4895-9927-90fa3254f860

In [ ]:
%%time
con.query('''
SELECT 
COUNT(distinct product_id) as product_cnt_30days,
COUNT(distinct user_id) as user_cnt_30days,
COUNT(distinct user_session) as user_session_cnt_30days
FROM '../Data/optimised_raw_data.parquet' WHERE event_type='view' ''').df()

CPU times: user 22.9 s, sys: 1.75 s, total: 24.7 s
Wall time: 27.9 s


product_cnt_30days  user_cnt_30days  user_session_cnt_30days
0              226209          4383932                 13757026

### Data Null Check on user_session

In [ ]:
con.query('''
SELECT COUNT(*) as null_cnt
FROM '../Data/optimised_raw_data.parquet'
WHERE event_type = 'view'
  AND (user_session IS NULL
       OR user_id IS NULL
       OR product_id IS NULL)''').df()

null_cnt
0         0

### Check how many distinct PDP view event is performed per session?

In [ ]:
con.query('''
SELECT 
user_session,
COUNT(distinct product_id) as count_unique_product_views_per_session
FROM '../Data/optimised_raw_data.parquet' 
WHERE event_type='view' 
GROUP BY 1''').df()

user_session  \
0         69b5d72f-fd6e-4fed-aa23-1286b2ca89a0   
1         7f596032-ccbf-4643-9bad-e36a209512b4   
2         d1e2f343-84bb-49bd-b13d-ca0f1ed9910e   
3         0444841c-38ef-410c-b11f-7b35ea4e5991   
4         31b7d4cf-dfac-4895-9927-90fa3254f860   
...                                        ...   
13757021  f467f7ab-6709-4836-80ba-0d1bd8d1a267   
13757022  b73ccf13-375c-4709-a0f3-d25cc6a11ae2   
13757023  5190c16b-f4f4-474c-ac75-de8ae413d82c   
13757024  7c8b6690-116e-4a3f-b573-be4f981393a8   
13757025  c923374f-27d5-4cd9-b705-26e341cd67cc   

          count_unique_product_views_per_session  
0                                              8  
1                                              4  
2                                              1  
3                                              1  
4                                              2  
...                                          ...  
13757021                                       3  
13757022                                       1  
13757023                                       1  
13757024                                       1  
13757025                                       1  

[13757026 rows x 2 columns]

In [ ]:
con.query('''
SELECT AVG(count_unique_product_views_per_session) as avg_unique_product_views_per_session  FROM (SELECT 
user_session,
COUNT(distinct product_id) as count_unique_product_views_per_session
FROM '../Data/optimised_raw_data.parquet' 
WHERE event_type='view' 
GROUP BY 1 ) as processed_table''').df()

avg_unique_product_views_per_session
0                              2.622409

#### In one user session on an average users visit 2 to 3 distinct PDPs.

### Count total events with non-null sessions

In [ ]:
con.query('''
SELECT COUNT(*) AS view_cnt
FROM '../Data/optimised_raw_data.parquet'
WHERE event_type = 'view'
  AND user_session IS NOT NULL''').df()

view_cnt
0  52490785

### EDA on user_sessions

In [ ]:
con.query('''
SELECT count(DISTINCT user_session) as user_session_cnt
FROM '../Data/optimised_raw_data.parquet' where event_type = 'view' ''').df()

user_session_cnt
0          13757026

In [ ]:
con.query('''
SELECT 
user_session,
COUNT(distinct product_id) as count_unique_product_views_per_session
FROM '../Data/optimised_raw_data.parquet' 
WHERE event_type='view' 
GROUP BY 1
''').df()

user_session  \
0         69b5d72f-fd6e-4fed-aa23-1286b2ca89a0   
1         7f596032-ccbf-4643-9bad-e36a209512b4   
2         d1e2f343-84bb-49bd-b13d-ca0f1ed9910e   
3         0444841c-38ef-410c-b11f-7b35ea4e5991   
4         31b7d4cf-dfac-4895-9927-90fa3254f860   
...                                        ...   
13757021  f467f7ab-6709-4836-80ba-0d1bd8d1a267   
13757022  b73ccf13-375c-4709-a0f3-d25cc6a11ae2   
13757023  5190c16b-f4f4-474c-ac75-de8ae413d82c   
13757024  7c8b6690-116e-4a3f-b573-be4f981393a8   
13757025  c923374f-27d5-4cd9-b705-26e341cd67cc   

          count_unique_product_views_per_session  
0                                              8  
1                                              4  
2                                              1  
3                                              1  
4                                              2  
...                                          ...  
13757021                                       3  
13757022                                       1  
13757023                                       1  
13757024                                       1  
13757025                                       1  

[13757026 rows x 2 columns]

In [ ]:
con.query('''
SELECT 
user_session,
COUNT(distinct product_id) as count_unique_product_views_per_session
FROM '../Data/optimised_raw_data.parquet' 
WHERE event_type='view' 
GROUP BY 1
HAVING COUNT(distinct product_id)>1
''').df()

user_session  \
0        69b5d72f-fd6e-4fed-aa23-1286b2ca89a0   
1        7f596032-ccbf-4643-9bad-e36a209512b4   
2        31b7d4cf-dfac-4895-9927-90fa3254f860   
3        c161400e-630b-4b59-8797-d9b9714444bf   
4        44ca9bc7-3ba8-4546-bdf5-10b2172ed883   
...                                       ...   
5340110  3df590c0-e667-4f5a-aa87-f47749cd6408   
5340111  e1252736-1d34-40c5-8aab-af2e5bff3201   
5340112  21f46782-fb36-4d5c-82d4-6dc26ae93fad   
5340113  4d614d05-3b0c-4504-a75e-9e09715f737f   
5340114  f467f7ab-6709-4836-80ba-0d1bd8d1a267   

         count_unique_product_views_per_session  
0                                             8  
1                                             4  
2                                             2  
3                                             2  
4                                             2  
...                                         ...  
5340110                                       2  
5340111                                       2  
5340112                                       2  
5340113                                       3  
5340114                                       3  

[5340115 rows x 2 columns]

## Data Analysis
* Number of entries for event_type = 'view': 52M+
* On an avg. users on Flipkart visit 2.6 (roughly 3) distinct PDPs per session
* In month of Jan2020 13M+ session happened on Flipkart where users visited atleast 1 PDP
* 5M+ session had more than 1 distinct PDP view